src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs

```diff
@@ -61,6 +61,15 @@ public async Task StartAsync(CancellationToken cancellationToken = default)
 
             foreach (IHostedService hostedService in _hostedServices)
             {
+                if (_applicationLifetime.ApplicationStopping.IsCancellationRequested)
+                {
+                    var startedServices = _hostedServices.TakeWhile(s => s != hostedService);
```

Comment:
I'd make some changes to this code:
- Don't call StopAsync from StartAsync, just return early from the loop.
- Change `_hostedServices` to represent the hosted services that successfully started instead of the entire list.
- Store the current list of hosted services in a local and enumerate that in start.

Reply:
I've updated the PR according to the suggestions you made 👍 By doing so, I changed a null check to check if `_hostedServices.Count != 0`, not sure if this check is necessary anymore 🤔 Looking forward hearing your thought on the latest changes!

Reply:
Seems like it would be wrong to fire NotifyStarted if we didn't successfully start but I'm not sure. It's weird to return from StartAsync without an error if we haven't successfully started.

Reply:
cc @eerhardt @maryamariyan for opinions

Reply:
I've been giving this some more thought, and here's my two cents: I think it is correct to return early without calling `NotifyStarted`. The reason why I think this makes sense it that the `IHostApplicationLifetime` would end up in a state where `ApplicationStopping` is triggered, but `ApplicationStarted` is not. To me, that signals that the application is not fully started, but is shutting down - which is also what is happening. I think there is an argument that this is a successful outcome of an graceful shutdown during start.

If, on the other hand, `StartAsync` would throw if the application is stopping before all hosted services are started, that means that there is no way to make a graceful stop when the host is in its start state. Is that desirable?

I appreciate that I do not have the picture here, and I'm looking forward hearing your thoughts on this!

Reply:
Yes I can see how this makes sense, my only concern is that we don't have an indication from the start call that it didn't complete. That might be surprising but maybe we can mitigate this with logging.  As a concrete example of what would happen when you use this, [these logs](https://github.com/dotnet/runtime/blob/09ad9e7eb457e5bc911b2292dc557dc02be78a20/src/libraries/Microsoft.Extensions.Hosting/src/Internal/ConsoleLifetime.cs#L69-L71) would never show up. Is that a fine? 

The other thing that could happen is that code would continue executing that is potentially not fully initialized. Code that assumes your hosted services ran and setup some initial state that is then used after the host has started. Maybe the code just needs to be robust against that?

```C#
public class Initializer : IHostedService
{
     private IMyCache _cache;
     public Initializer(IMyCache cache) 
     {
          _cache = cache;
     }
     public Task StartAsync() => _cache["expensiveWork"] = ExpensiveCalculation();
}
```

Then in Main:

```C#
static void Main(string[] args)
{
    using var host = CreateDefaultBuilder(args).Build();
    host.Start();
    var cache = host.Services.GetRequiredService<IMyCache>();
    // Who needs defensive coding
    Use(cache.Cache["expensiveWork"].Value); // This will go boom
}
```

Reply:
I see what you mean, and understand your concerns 👌 Following the same logical argument as earlier, I think it is correct not to log _Application started. Press Ctrl+C to shut down._, since the application did not transit from starting to started state. I definitely think this can be made clearer by adding logging, as you suggest.

A few thoughts on your cache example:

1. I don't think it is unreasonable to ask the developer to check the state of the host (using `IHostApplicationLifetime`) here. Correct me if I'm wrong, but it is the one source of truth for the host's state. I understand that what I'm suggesting is a changed behavior. It would be interesting to know how common of a scenario it is to have code like that in combination with a call to `StopApplication` made in a hosted service. 
2. If we would throw on `_applicationLifetime.ApplicationStopping` instead of returning, the behavior changes from how it works today, as it is currently possible that all hosted services starts even though the application is instructed to stop.
3. A wild idea, that probably is not feasible:  It is possible to decorate the host's `IServiceProvider` so that if services are resolved from a not fully started host, it logs a warning before resolving the service, or perhaps it should not be possible to resolve services if the start is interrupted? This is probably overkill, and the logging from the incomplete start would probably suffice - just widening the solution space 😆